compiler/hash-ast/src/operator.rs

```diff
@@ -1,37 +1,201 @@
-#[derive(Debug)]
-pub enum CompoundFn {
-    Leq,
-    Geq,
-    Lt,
-    Gt,
+/// This implementation will be used for error messages. The debug formatter is
+/// mainly used in error reports whereas the display formatter is used when
+/// printing code from tokens.
+impl std::fmt::Debug for OperatorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OperatorKind::EqEq => write!(f, "eq"),
+            OperatorKind::NotEq => write!(f, "neq"),
+            OperatorKind::BitOr => write!(f, "bit_or"),
+            OperatorKind::Or => write!(f, "orl"),
+            OperatorKind::BitAnd => write!(f, "bit_and"),
+            OperatorKind::And => write!(f, "and"),
+            OperatorKind::BitXor => write!(f, "bit_xor"),
+            OperatorKind::Exp => write!(f, "exp"),
+            OperatorKind::Gt => write!(f, "gt"),
+            OperatorKind::GtEq => write!(f, "gt_eq"),
+            OperatorKind::Lt => write!(f, "lt"),
+            OperatorKind::LtEq => write!(f, "lt_eq"),
+            OperatorKind::Shr => write!(f, "shr"),
+            OperatorKind::Shl => write!(f, "shl"),
+            OperatorKind::Add => write!(f, "add"),
+            OperatorKind::Sub => write!(f, "sub"),
+            OperatorKind::Mul => write!(f, "mul"),
+            OperatorKind::Div => write!(f, "div"),
+            OperatorKind::Mod => write!(f, "mod"),
+            OperatorKind::As => write!(f, "as"),
+        }
+    }
+}
+
+impl OperatorKind {
+    /// This returns if an operator is actually re-assignable. By re-assignable, this is in the sense
+    /// that you can add a '=' to mean that you are performing a re-assigning operation using the left
+    /// hand-side expression as a starting point and the rhs as the other argument to the operator.
+    /// For example, `a += b` is re-assigning because it means `a = a + b`.
+    pub fn is_re_assignable(&self) -> bool {
+        matches!(
+            self,
+            OperatorKind::BitOr
+                | OperatorKind::Or
+                | OperatorKind::BitAnd
+                | OperatorKind::And
+                | OperatorKind::BitXor
+                | OperatorKind::Exp
+                | OperatorKind::Shr
+                | OperatorKind::Shl
+                | OperatorKind::Add
+                | OperatorKind::Sub
+                | OperatorKind::Mul
+                | OperatorKind::Div
+                | OperatorKind::Mod
+        )
+    }
+
+    /// operators exhibit short circuiting behaviour.
+    pub fn is_lazy(&self) -> bool {
+        matches!(self, OperatorKind::And | OperatorKind::Or,)
+    }
+
+    /// Compound functions that use multiple function calls when they are transformed.
+    pub fn is_compound(&self) -> bool {
+        matches!(
+            self,
+            OperatorKind::Lt | OperatorKind::Gt | OperatorKind::LtEq | OperatorKind::GtEq
+        )
+    }
+
+    pub fn to_string(&self) -> &'static str {
```

Comment:
Possibly better to call this `to_str` because `to_string` is also a trait method of ToString/Display and since this type implements Display that can get confusing